### DISCUSSION TITLE
Nuxt3 and Pinia how to get localstorage items

### DISCUSSION THREAD

**Discussion by Moemen02** [nuxt/nuxt#15996] [Questions] [answered]
I'm using `Nuxt3` and `pinia`, and I am pretty new to them both. I am trying to get a token from the localstorage. Now I know that you can't do that because nuxt is SSR and it does not have a way to get the client's localstorage. So the localstorage has to be defined after the components got loaded. If I understand that correctly.

So in my website when a user logges in the token is getting successfully stored in the localstorage, but it I can't access it when I reload the page (only if I use the onMounted method).

When I use the onMounted method the I can get the token successfully. But then I have another problem. I want to add a "isAuthenticated" middleware, so when you are already logged in you can't go to the login page again. So I am trying to get the token from the localstorage, and when I do that I am getting either an empty string or "localstorage" is not defined

The login function `<script setup>` (this is working fine) to upload the token in the localStorage

````
import {useTokenStore} from "../store/user/token";
const tokenStore = useTokenStore()

async function login() {
  await useLazyFetch('/api/auth/login', {
    method: 'POST',
    body: {
      email: `${inpEmail}`,
      password: `${inpPassword}`
    }
  })
  .then((response) => {
    responseToken = response.data.value
    if (responseToken!=="fout"){
      for (let token in responseToken){
        tokenStore.storeToken(responseToken[token])
      }
    } else {
      errorMsg.value = true
    }
  })
  .catch((error) => {
    console.log(error.data)
  })
}
````

This function (also in <script setup> is how I am getting (for now) the token from the localstorage. As you can see I am getting the token after my component is loaded, and restoring it in my state

````
onMounted(() => {
  tokenStore.storeToken(localStorage.getItem('token'))
})`

This is my my store/user/token.js. Here I set the token in the storage

`import { defineStore } from "pinia";

export const useTokenStore = defineStore("token", {
    state: () => ({
        userToken: process.server ? '' : localStorage.getItem('token')
    }),
    getters: {
        getUserToken: state => {
            console.log(process.server)
            return state.userToken
        }
    },
    actions: {
        storeToken(token){
            this.userToken = token
            localStorage.setItem('token', this.userToken)
        }
    }
})
````

So after the code above I tried to get in my middleware the token from my localstorage in my `middleware/auth.js:`
````
export default defineNuxtRouteMiddleware((to, from) =>{
    const useToken = useTokenStore()
    console.log(useToken.getUserToken + " token")
    console.log(localStorage.getItem('token'))
)}
````

So when I run this with the first console.log im getting an empty string. When I try to run it with the second console.log I'm getting this error:

````
500
localStorage is not defined

at fn (./node_modules/nuxt/dist/app/nuxt.mjs:151:27)
at Object.callAsync (./node_modules/unctx/dist/index.mjs:49:19)
at Module.callWithNuxt (./node_modules/nuxt/dist/app/nuxt.mjs:153:23)
at ./node_modules/nuxt/dist/pages/runtime/router.mjs:110:50
at runMicrotasks ()
at processTicksAndRejections (node:internal/process/task_queues:96:5)
````

So basically how can I get my token, so I can use it in my middleware to check if my token exist?

I know that I just can turn SSR off, but I want it on so my website is better for SEO

**Comment by IsraelOrtuno** [comment, marked as answer]
You could store your token in a cookie (see https://nuxt.com/docs/api/composables/use-cookie#usecookie) so you will have access to your token in both client and server side.

Also check this which may be helpful for creating "stateful cookies" https://github.com/nuxt/framework/issues/2416#issuecomment-1315223979

**Reply by Moemen02** [reply]
cookies are working fine thank you!

**Reply by sribadal** [reply]
Could you maybe share your cookie solution?

**Reply by Moemen02** [reply]
@shivam-badal here is how I store my token with pinia in the cookies
```
state: () => ({
        userToken:useCookie('token')
    }),
    getters: {
        getUserToken: state => {
            return state.userToken
        }
    },
    actions: {
        storeToken(token){
            this.userToken = token
            const newCookie = useCookie('token', {
                maxAge: 60*24*28,
                sameSite: true,
                secure: true,
            })
            newCookie.value = this.userToken
        },
    }
```

**Reply by sribadal** [reply]
Thx man, had a solution, but yours seems more elegant

**Comment by lkgit1** [comment]
If you still prefer to make use of localStorage instead of cookies... check this simple guide I made, on Medium: https://medium.com/@lk4real/working-with-localstorage-in-nuxt-3-c27618ab89bd
